src/main/java/org/jabref/gui/fieldeditors/LinkedFileViewModel.java

```diff
@@ -415,6 +416,7 @@ public void download() {
                 LinkedFile newLinkedFile = LinkedFilesEditorViewModel.fromFile(destination, databaseContext.getFileDirectoriesAsPaths(filePreferences), externalFileTypes);
                 linkedFile.setLink(newLinkedFile.getLink());
                 linkedFile.setFileType(newLinkedFile.getFileType());
+                entry.addFile(0, newLinkedFile);
```

Comment:
@btut Thanks for your implementation. I fear this addition here now duplicates the file if you right-click an online link and select "Download". Can you please double check this.
(In case this is really an issue, one option would be to use `prepareDownloadTask` below in the importer. Then you can also display a nice progress window for the downloads.)

Reply:
Good point @tobiasdiez. I checked, and this does indeed duplicate the entry.
the progress-bar next to the link entry does look very nice indeed and would probably be good to have when working with large files or slow connections. I can imagine people thinking it didn't work because the file doesn't show up immediately.

I don't understand your suggestion of using prepareDownloadTask, though. Isn't that implicitly used by the download() method anyway?

As far as I can tell the right-click -> download menu item just calls the download-function of the LinkedFileViewModel. In that case, the LInkedFileViewModel is the one that is actually displayed. Therefore the progress bar can be shown. However; I have no access to that object, correct? I have the LinkedFileViewModel that I use for the download, but I cannot add it to any LinkedFilesEditorViewModel to be displayed.

One quick fix I could think of would be to keep the online-link even for right-click->download actions or delete the link for both methods. There would still be no progress bar for the imported files, but at least it stops duplication.

The best fix (IMO) would be to gain somehow access to the actual LinkedFileViewModel of the online link after the import and call download from there, basically imitating a user right-clicking the item and selecting download. I would need some pointers on how to get that access though.

Reply:
I guess it would work if I moved the downloadProgressProperty from LinkedFileViewModel to LinkedFile. That object lives in the BibEntry and (AFAICT) and would be the same both for the importer and the LinkedFileViewModel.
Would moving that property be an acceptable option?

Reply:
@btut Have a look at the method `addLinkedFileFromURL` in DownloadFullTextAction, that solves a similar case 
In the importer you would create a progressbar/dialog and bind it to the prepareDownloadTask.progressProperty

Reply:
That method just notifies the user after a successful download. I see no progress bar. My issue is not in getting the progress of the download, but where to display the progress bar. Just popping up a dialogue with a bar is no viable solution in my opinion. When I import an entry I don't want to watch a progress bar while most of the entry is already there, I want to start working with the entry.

As I said, a progress bar would be very helpful, but I don't think popping a dialogue (or remaining in the import dialogue as long as the files are not ready, if that is what you mean) would be a good idea. I like how the right-click -> download action puts a small progress bar in the file entry itself. I can work with the bib entry, and if I try to open up the file, I see the progress bar.

However, since there is no access to the linked files editor from the importer, I cannot add the progress bar there. I think moving the download-related properties to LinkedFile would work quite well; I just don't know if it is compliant with how JabRef is developed.

One more thought on that addLinkedFileFromURL method you mentioned. It seems to share a lot of code with the download method of LinkedFileViewModel. If we move the properties I was talking about to the LinkedFile class; the LinkedFilesEditorViewModel would always correctly display the progress no matter where the download() method of LinkedFileViewModel would be called from. In that case, I would suggest calling the download Method from addLinkedFileFromURL to reduce code duplication and have a unique way of handling file downloads and showing it's progress.

Reply:
Because it is a small change I just wanted to try it out and unfortunately it does not seem to work. You can check it out in [my fork](https://github.com/btut/jabref/tree/feature/downloadFilesOnImport), I guess once this pull request is reopened due to the file-duplication issue this would show up here as well.

The right-click -> download action still works fine and shows the progress. The file is now not duplicated, but the original online link is kept.

As I said, the progress for the imported files still does not show. I added some log messages in the LinkedFile constructor and noticed that after clicking import, it is called twice. I guess the one showed is not the one I bind the download-properties and therefore the progress is not shown. Why are there multiple LinkedFile objects for the same file?

Reply:
To me this looks like the LinkedFile object from the Import Dialogue is not the same LinkedFile object that is then displayed, can somebody confirm?

Reply:
I will take a look at it later

Reply:
The latest commit in my branch now deletes the online file after the download. As @Siedlerchr said removing the online link is not a good idea but the right-click -> download action did that all along, I guess it is best to create a preference for this after all.

This also works only when downloaded using the right-click -> download action. If the file is downloaded using the import dialogue, the online link is not deleted.

Reply:
@Siedlerchr thank you!
I guess the progress bar is not really that important though. We can go the addLinkedFileFromURL way and just display a message once done.
I just found the progress bar very neat.

Reply:
Thanks @btut for having a second look at this problem. 

Moving the download progress properties to the `LinkedFile` class feels a bit inconvenient. This class should represent a real linked file (or web link) while the download action is something related to an UI action which is better placed in the `LinkedFileViewModel`  (which is the UI-wrapper of `LinkedFile`).

As Christoph said, I would do it similar to the following:
https://github.com/JabRef/jabref/blob/3d24714deb07978972b921025551d5328faff039/src/main/java/org/jabref/gui/externalfiles/DownloadFullTextAction.java#L159-L168
The progress can be displayed by disabling the "Import" button but not yet closing the dialog as in the 
https://github.com/JabRef/jabref/blob/master/src/main/java/org/jabref/gui/EntryTypeView.java#L76
You said you don't like to use the dialag to show the progress. Why is this? If the dialog is closed but the download is still going on in the background, than we have to handle a couple of situations that are complex. E.g. user closes JabRef before all downloads are done, user removes online link before they are downloaded, etc.

Reply:
Hi,

> This class should represent a real linked file (or web link) while the download action is something related to a UI action which is better placed in the LinkedFileViewModel (which is the UI-wrapper of LinkedFile).

As I said, I am not very familiar with GUI programming so I have no idea how such things are usually done. From an outsiders view though it seems logical to me to place the download-progress properties in the linked files. As I see it, they are more related to the status of the actual file than with the UI action and therefore belong with the model. Just my thoughts though and since it does not fix anything I don't really care at all.

> You said you don't like to use the dialogue to show the progress. Why is this?

It just seems disruptive to me. If I had to choose between waiting in the import dialogue and having no progress bar, I would definitely choose the second option. I like how these background tasks are handled in JabRef and I think it would be a shame to have a background task which blocks the UI.

> If the dialogue is closed but the download is still going on in the background then we have to handle a couple of situations that are complex.

Isn't that already the case with the right-click -> download action though?

> E.g. user closes JabRef before all downloads are done

I didn't try this out, but I think the current implementation would then leave the bib-entry as it was, as it is only changed after the download is finished. The file would be corrupt but still there and the user might link it manually the next time as it already has the correct name and JabRef would pick it up easily. This could be a problem.

> user removes online link before they are downloaded, etc

I think this would be fine. The download would still go on and when it's done the new entry is added (as my code does not overwrite the online-link entry but add a new one). If one would then close JabRef during the download, the next time there would be neither an online-link nor a file entry, which I guess would be fine as the user explicitly deleted the link.

I noticed in the code (never while running JabRef and I cannot find any piece of code where it is actually used and even setting it's visibility to true does not show anything, maybe it's just dead code) that there is a progress bar in the right end of the status line at the bottom of the frame. What about using that one? The user is not disrupted in his flow and upon closing, we could pop a dialogue warning about ongoing downloads.

If that progress bar is dead code, I guess there was one at some point which was then removed, so you probably didn't like it. In that case, what do you think about having a small loading circle or a download icon in the top right, next to the menu bar? It could be turning/greyed out while downloading and a tooltip could display the progress of single files.

Reply:
Ok, let's continue trying to do it in the background. If users experience problems, then we can investigate of how to best solve them and maybe go back to the blocking-ui strategy if nothing else works.

> I noticed in the code (never while running JabRef and I cannot find any piece of code where it is actually used and even setting it's visibility to true does not show anything, maybe it's just dead code) that there is a progress bar in the right end of the status line

The old JabRef 3.x contained a status bar, which was removed since it wasn't used that much. So if there is still code related to this in the code base, then it is definitely dead code and can safely be removed. May I ask you to open a PR removing it? Thanks!

> If that progress bar is dead code, I guess there was one at some point which was then removed, so you probably didn't like it. In that case, what do you think about having a small loading circle or a download icon in the top right, next to the menu bar? It could be turning/greyed out while downloading and a tooltip could display the progress of single files.

I like it! Very good idea. Might also be helpful to show progress for other actions as well (e.g. cleanup). One possibility to implement this would be adding an ObservableList of open tasks (including their progress) in `StatusManager`.

I guess the most pressing thing would be to fix the download right-click, and then afterwards worry about a nice progress display.

Reply:
> I like it! Very good idea. Might also be helpful to show progress for other actions as well (e.g. cleanup). One possibility to implement this would be adding an ObservableList of open tasks (including their progress) in StatusManager.

That was the basic idea. BackgroundTask already has a workDonePercentageProperty, so all thats left to do is build a list of them and display it.
I have a couple of questions though:

- The list of tasks would consist of a text label and a progress bar. BackgroundTask has a StringProperty message, I would use that one for the label. Unfortunately, the message does not seem to be set by the file downloaders. It is only set by the UnlinkedFilesCrawler. Did I miss something? If not, I would add a message (something like "Downloading file from <url>"). For other tasks that use BackgroundTask we could either add a message parameter to the Background-Task constructor to force users to provide a sensible message, or we could set the default message to "<class name> in progress", so at the very least it says "BackgroundTask" in progress, for child-classes it would take the child's name and produce a somewhat usable label.

- I can't seem to find the fxml file for the main JabRef frame where I would add the status wheel. Could you please help me out here?

Reply:
@btut  The StringProperty is as the name implies a property that you can set for the task.In the File Download task you can simply set it to (calling setValue) on it with a text. e.g. "Downloading files".
As this property is exposed you can bind a label to it. 
For a quick understanding what those properties are and how they work:
https://www.dummies.com/programming/java/javafx-binding-properties/

JabRefFrame has no fxml. Have a look at the init/ initLayout Method().
You can always add

Reply:
Thanks for the input. I already started drafting some code and came by the properties and did some reading.
I implemented a simple dialog using the TaskProgressView from javafx control. That looks pretty nice IMO. Unfortunately (and quite obviously) it works on a list of javafx tasks, not JabRef BackgroundTasks.
I noticed that BackgroundTasks are converted to javafx tasks anyways, so when that is done I enlist them in an ObservableList I created in StateManager. The tasks show up, but the bindings do not seem to work, I still have to figure out why. So right now, one can see a list of tasks, but the tasks have no title, no message and an infinite progress bar.

If you don't mind I would like to open a WIP PR and if I need some help I would start a discussion there. Is that ok?

Reply:
Yes, go ahead with a PR, makes it easier to help for specific questions